c2rust-analyze/src/c_void_casts.rs

```diff
@@ -0,0 +1,484 @@
+    /// but currently rustc always uses a temporary for the LHS of the
+    /// cast in the desired free(ptr as *mut c_void) pattern, so I think we
+    /// can conservatively treat the Place as unmodified only if (1) it's
+    /// just a Local, with no projections, and (2) that local isn't
+    /// mentioned in the LHS or RHS of any statements between the cast and
+    /// the call.
+    fn is_place_modified_by_statement(p: &Place, stmt: &Statement) -> bool {
+        if !p.projection.is_empty() {
+            return false;
+        }
+
+        let local = p.local;
+
+        if let StatementKind::Assign(assign) = &stmt.kind {
+            let (lhs, rv) = &**assign;
+            if lhs.local == local {
+                return true;
+            }
+
+            use Rvalue::*;
+            let rv_place = match rv {
+                Use(op) => op.place(),
+                Repeat(op, _) => op.place(),
+                Ref(_, _, p) => Some(*p),
+                ThreadLocalRef(..) => None,
+                AddressOf(_, p) => Some(*p),
+                Len(p) => Some(*p),
+                Cast(_, op, _) => op.place(),
+                BinaryOp(..) => None,
+                CheckedBinaryOp(..) => None,
+                NullaryOp(..) => None,
+                UnaryOp(_, op) => op.place(),
+                Discriminant(p) => Some(*p),
+                Aggregate(..) => None,
+                ShallowInitBox(op, _) => op.place(),
+                CopyForDeref(..) => None,
+            };
+
+            if let Some(rv_local) = rv_place.map(|p| p.local) {
+                return local == rv_local;
+            }
+        }
+
+        false
+    }
+
+    /// Search for the first [Rvalue::Cast] from a void pointer
+    /// in a sequence of [Statement]s. We expect that the
+    /// cast is the first non-[StatementKind::StorageDead]
+    /// statement in the block, a special case for
+    /// [Terminator]s whose destination is casted from a
+    /// void pointer to some other pointer type.
+    fn find_first_cast(
+        statements: &[Statement<'tcx>],
+        c_void_ptr: CVoidPtr<'tcx>,
+    ) -> Option<(usize, CVoidCast<'tcx>)> {
+        statements
+            .iter()
+            .enumerate()
+            .skip_while(|(_, stmt)| matches!(stmt.kind, StatementKind::StorageDead(_)))
+            .find_map(|(i, stmt)| {
+                Some((
+                    i,
+                    c_void_ptr.get_cast_from_stmt(CVoidCastDirection::From, stmt)?,
+                ))
+            })
+    }
+
+    /// Search for the last cast to a void pointer in a sequence of
+    /// [Statement]s.
+    fn find_last_cast(
+        statements: &[Statement<'tcx>],
+        c_void_ptr: CVoidPtr<'tcx>,
+    ) -> Option<(usize, CVoidCast<'tcx>)> {
+        // Get the last cast in the current block, and ensure that the destination
+        // of that cast remains unmodified between the cast and the call in which
+        // the destination place is used.
```

Comment:
Can you move the comment up to part of the doc comment?

Reply:
ace07810